cmd/gen-ca/main.go

```diff
@@ -0,0 +1,363 @@
+		}
+	default:
+		return nil, errors.New("unsupported key type")
+	}
+
+	return &x509Signer{
+		ctx:          ctx,
+		session:      session,
+		objectHandle: privateHandle,
+		keyType:      keyType,
+		pub:          pub,
+	}, nil
+}
+
+var algToString = map[string]x509.SignatureAlgorithm{
+	"SHA256WithRSA":   x509.SHA256WithRSA,
+	"SHA384WithRSA":   x509.SHA384WithRSA,
+	"SHA512WithRSA":   x509.SHA512WithRSA,
+	"ECDSAWithSHA256": x509.ECDSAWithSHA256,
+	"ECDSAWithSHA384": x509.ECDSAWithSHA384,
+	"ECDSAWithSHA512": x509.ECDSAWithSHA512,
+}
+
+type certProfile struct {
+	SignatureAlgorithm string
+
+	Subject struct {
+		CommonName   string
+		Organization string
+		Country      string
+	}
+
+	NotBefore string
+	NotAfter  string
+
+	OCSPURL   string
+	CRLURL    string
+	IssuerURL string
+
+	PolicyOIDs []string
+}
+
+func parseOID(oidStr string) (asn1.ObjectIdentifier, error) {
+	var oid asn1.ObjectIdentifier
+	for _, a := range strings.Split(oidStr, ".") {
+		i, err := strconv.Atoi(a)
+		if err != nil {
+			return nil, err
+		}
+		oid = append(oid, i)
+	}
+	return oid, nil
+}
+
+const dateLayout = "2006-01-02 15:04:05"
+const serialLength = 16
+
+func constructCert(ctx pkcs11helpers.PKCtx, profile *certProfile, pubKey []byte, session pkcs11.SessionHandle) (*x509.Certificate, error) {
+	notBefore, err := time.Parse(dateLayout, profile.NotBefore)
+	if err != nil {
+		return nil, err
+	}
+	notAfter, err := time.Parse(dateLayout, profile.NotAfter)
+	if err != nil {
+		return nil, err
+	}
+
+	var policyOIDs []asn1.ObjectIdentifier
+	for _, oidStr := range profile.PolicyOIDs {
+		oid, err := parseOID(oidStr)
+		if err != nil {
+			return nil, err
+		}
+		policyOIDs = append(policyOIDs, oid)
+	}
+
+	sigAlg, ok := algToString[profile.SignatureAlgorithm]
+	if !ok {
+		return nil, fmt.Errorf("unsupported signature algorithm %q", profile.SignatureAlgorithm)
+	}
+	// generate serial number
+	serial, err := ctx.GenerateRandom(session, serialLength)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate serial number: %s", err)
+	}
+	keyHash := sha256.Sum256(pubKey)
+
+	cert := &x509.Certificate{
+		SignatureAlgorithm:    sigAlg,
+		SerialNumber:          big.NewInt(0).SetBytes(serial),
+		BasicConstraintsValid: true,
+		IsCA: true,
+		Subject: pkix.Name{
+			CommonName:   profile.Subject.CommonName,
+			Organization: []string{profile.Subject.Organization},
+			Country:      []string{profile.Subject.Country},
+		},
+		NotBefore:             notBefore,
+		NotAfter:              notAfter,
+		OCSPServer:            []string{profile.OCSPURL},
+		CRLDistributionPoints: []string{profile.CRLURL},
+		IssuingCertificateURL: []string{profile.IssuerURL},
+		PolicyIdentifiers:     policyOIDs,
+		KeyUsage:              x509.KeyUsageCertSign & x509.KeyUsageCRLSign,
+		SubjectKeyId:          keyHash[:],
+	}
+
+	return cert, nil
+}
+
+func main() {
+	module := flag.String("module", "", "PKCS#11 module to use")
```

Comment:
Is there a default for this that we can set for the Linux distro we expect to use this on?

Reply:
The PKCS#11 module you use is dependent on the hardware/software you are communicating with. I think putting any defaults here would leak more information about things we use than we are probably comfortable with.